casper/src/main/scala/coop/rchain/casper/engine/LfsBlockRequester.scala

```diff
@@ -192,7 +198,16 @@ object LfsBlockRequester {
         val blockNumber = ProtoUtil.blockNumber(block)
         for {
           // Mark block as received and calculate minimum height (if latest)
-          receivedResult <- st.modify(_.received(block.blockHash, blockNumber))
+          receivedResult <- st.modify(st => {
+                             val isAbLatest = latestMessages.contains(block.blockHash)
+                             if (isAbLatest) {
+                               val selfJustification =
+                                 ProtoUtil.creatorJustification(block).map(_.latestBlockHash)
+                               // if message received is latest as per approved block - add its self justification
+                               // to target latest messages that has to be pulled
+                               st.received(block.blockHash, blockNumber, selfJustification)
+                             } else st.received(block.blockHash, blockNumber, None)
```

Comment:
These lines are duplicated in first two arguments. If block should only be for third argument.